File: library/src/test/java/de/westnordost/countryboundaries/CountryAreasTest.java
package de.westnordost.countryboundaries;

import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class CountryAreasTest
{

	private static final Point[] BIG_SQUARE = new Point[]{p(0, 0), p(0, 10), p(10, 10), p(10, 0)};

	private static final Point[] HOLE = new Point[]{p(2, 2), p(2, 8), p(8, 8), p(8, 2)};

	private static final Point[] SMALL_SQUARE = new Point[]{p(4, 4), p(4, 6), p(6, 6), p(6, 4)};


	@Test public void coversSimplePolygon()
	{
		assertTrue(new CountryAreas("A",
				new Point[][]{BIG_SQUARE},
				new Point[][]{}).covers(p(5,5))
		);
	}

	@Test public void doesNotCoverHole()
	{
		assertFalse(new CountryAreas("A",
				new Point[][]{BIG_SQUARE},
				new Point[][]{HOLE}).covers(p(5,5))
		);
	}

	@Test public void doesCoverPolygonInHole()
	{
		assertTrue(new CountryAreas("A",
				new Point[][]{BIG_SQUARE , SMALL_SQUARE},
				new Point[][]{HOLE}).covers(p(5,5))
		);
	}

	private static Point p(double x, double y)
	{
		return new Point(Fixed1E7.doubleToFixed(x),Fixed1E7.doubleToFixed(y));
	}
}

File: library/src/commonTest/kotlin/de/westnordost/countryboundaries/CountryAreasTest.kt
package de.westnordost.countryboundaries

import kotlin.test.*

internal class AreasTest {
    @Test fun covers_simple_polygon() {
        assertTrue(CountryAreas("A", listOf(BIG_SQUARE), listOf()).covers(p(5, 5)))
    }

    @Test fun does_not_cover_hole() {
        assertFalse(CountryAreas("A", listOf(BIG_SQUARE), listOf(HOLE)).covers(p(5, 5)))
    }

    @Test fun does_cover_polygon_in_hole() {
        assertTrue(CountryAreas("A", listOf(BIG_SQUARE, SMALL_SQUARE), listOf(HOLE)).covers(p(5, 5)))
    }

    @Test fun only_upper_left_edge_counts_as_inside() {
        val areas = CountryAreas("A", listOf(BIG_SQUARE), listOf())

        assertTrue(areas.covers(p(0, 0)))
        assertTrue(areas.covers(p(5, 0)))
        assertTrue(areas.covers(p(0, 5)))
        assertFalse(areas.covers(p(0, 10)))
        assertFalse(areas.covers(p(10, 0)))
        assertFalse(areas.covers(p(5, 10)))
        assertFalse(areas.covers(p(10, 5)))
        assertFalse(areas.covers(p(10, 10)))
    }
}

private val BIG_SQUARE = listOf(p(0, 0), p(0, 10), p(10, 10), p(10, 0))

private val HOLE = listOf(p(2, 2), p(2, 8), p(8, 8), p(8, 2))

private val SMALL_SQUARE = listOf(p(4, 4), p(4, 6), p(6, 6), p(6, 4))